core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -52,17 +52,43 @@
     private final DictionaryId dictionarySourceId;
     private final boolean mayHaveNull;
 
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids)
+    public static Block create(int positionCount, Block dictionary, int[] ids)
     {
-        this(0, positionCount, dictionary, ids, false, false, randomDictionaryId());
+        return createInternal(positionCount, dictionary, ids, randomDictionaryId());
     }
 
     /**
      * This should not only be used when creating a projection of another dictionary block.
      */
-    public DictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
+    public static Block createProjectedDictionaryBlock(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
     {
-        this(0, positionCount, dictionary, ids, false, false, dictionarySourceId);
+        return createInternal(positionCount, dictionary, ids, dictionarySourceId);
+    }
+
+    private static Block createInternal(int positionCount, Block dictionary, int[] ids, DictionaryId dictionarySourceId)
+    {
+        if (positionCount == 0) {
+            return dictionary.copyRegion(0, 0);
+        }
+        if (positionCount == 1) {
+            return dictionary.getRegion(ids[0], 1);
+        }
+
+        // if dictionary is an RLE then this can just be a new RLE
+        if (dictionary instanceof RunLengthEncodedBlock rle) {
+            return new RunLengthEncodedBlock(rle.getValue(), positionCount);
+        }
+
+        // unwrap dictionary in dictionary
+        if (dictionary instanceof DictionaryBlock dictionaryBlock) {
+            int[] newIds = new int[positionCount];
```

Comment:
Unnesting is not neccecerly always beneficial without looking at context, e.g: consider join:
```
left_col1 | left_col2 | right_col1
===================================
    row42 |     row42 |     right1
    row42 |     row42 |     right2
    ...
    row42 |     row42 |     rightN
```
`row42` from probe is repeated N times. Right not in join we will use dictionary (`getPosition`) to avoid copying `row42` N times. This means that `dictionaryId` for block `left_col1, left_col2` can be same.

If there is now dictionary aware operator on `left_col1, left_col2`, then because `left_col1, left_col2` have same `dictionaryId` we can process it once rather than `N` times.

However, if you unnest dictionary always, then you have to drop common `dictionaryId` for `left_col1, left_col2` in this method.

Reply:
This seems like a very rare scenario compared to the benefits due to reduced complexity and being able to avoid megamorphic calls in certain places (all part of the effort tracked under https://github.com/trinodb/trino/issues/14237)